ext/json/json_encoder.c

```diff
@@ -104,7 +104,11 @@ static inline void php_json_encode_double(smart_str *buf, double d, int options)
 {
 	size_t len;
 	char num[PHP_JSON_DOUBLE_MAX_LENGTH];
-	php_gcvt(d, (int)EG(precision), '.', 'e', &num[0]);
+	if (JSON_G(precision) == -1) {
+		php_0cvt(d, 17, '.', 'e', num);
+	} else {
+		php_gcvt(d, JSON_G(precision), '.', 'e', num);
+	}
```

Comment:
Actually I was wrong and didn't look to php_gcvt properly (I just check zend_dtoa). It uses ndigits for selecting if the number should be printed in E-style so it's important:

https://github.com/php/php-src/blob/f6d196a38f484603452ca0ffaa241940c452ac99/main/snprintf.c#L163

I'm not sure that we should ignore that and also not sure if (-1) is a good way how to change mode because it means disallowing precision setting (selecting when the E-style is used). Maybe something like ini "serialization_precision_mode" would be better but it's maybe too many ini settings...